Convert Agg extension to pybind11

QuLogic opened the pull request:
## PR summary

This is somewhat partially implemented, in that the NumPy array parts are not changed yet. I think I'll want to do path first, to make this simpler. I'll leave this as draft until then.

## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [n/a] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [n/a] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

QuLogic commented:
I rebased this on top of #27087 so that I can re-use the type casters that it has. However, there is still an issue with the `GCAgg` type caster, but I'll leave that for when the other PR is in.

ianthomas23 commented:
@QuLogic When you have time and motivation to come back to this, ping me and I will review it.

QuLogic commented on src/py_converters_11.h:
```diff
@@ -113,14 +177,14 @@ namespace PYBIND11_NAMESPACE { namespace detail {
             auto simplify_threshold = src.attr("simplify_threshold").cast<double>();
 
-            if (!value.set(vertices.ptr(), codes.ptr(),
+            if (!value.set(vertices.inc_ref().ptr(), codes.inc_ref().ptr(),
```
I don't know why this change is needed now, but otherwise things randomly crash.

tacaswell replied:
We think there are decrements in the iterator destructor there is a decr which we think we need to keep as there are still other usages of the iterators from c++ that we need to remove first.

QuLogic commented on src/py_converters_11.h:
```diff
@@ -113,14 +177,14 @@ namespace PYBIND11_NAMESPACE { namespace detail {
+            if (!value.set(vertices.inc_ref().ptr(), codes.inc_ref().ptr(),
                            should_simplify, simplify_threshold)) {
-                return false;
+                throw py::error_already_set();
```
This was technically wrong before, but otherwise, `pybind11` would substitute its own generic "cannot convert type PyType to C++Type" error message, so not fatal.

QuLogic commented on src/py_converters_11.h:
```diff
@@ -146,6 +268,32 @@ namespace PYBIND11_NAMESPACE { namespace detail {
+            value.join = src.attr("_joinstyle").cast<agg::line_join_e>();
+            value.dashes = src.attr("get_dashes")().cast<Dashes>();
+            value.cliprect = src.attr("_cliprect").cast<agg::rect_d>();
+            /* value.clippath = src.attr("get_clip_path")().cast<ClipPath>(); */
+            convert_clippath(src.attr("get_clip_path")().ptr(), &value.clippath);
```
If I leave the casting version in, then all sorts of memory corruption occurs, and I've yet to figure out why. Instead of wasting even more time on that now, I'd rather get this going, and once we drop NumPy headers, come back to clean this up properly.

tacaswell replied:
ok

QuLogic commented:
I rebased and added the usual `namespace py = pybind11;`; this was left out earlier due to conflicts and has been fixed by other cleanup PRs (though I don't remember which now). I also converted the Agg-only `array_view` usage to pybind11; the ones shared with `_path` will be left for a later PR (or it could be done here too, but this is large enough, I think?)
